### ISSUE TITLE
sourceBreakdown service shows values that are too high

### ISSUE THREAD

**Issue by nickdos** [AtlasOfLivingAustralia/logger-service#31]
E.g. https://logger-test.ala.org.au/service/sourceBreakdown?eventId=1002

shows:
```json
{
  "thisMonth": {
    "events": 493,
    "records": 396882
  }
}
```
doing a single download (1 event and 31 records) it changes to:
```json
{
  "thisMonth": {
    "events": 510,
    "records": 397006
  }
}
```

It seems the `event` count goes up by a factor of `17` and the `records` count by a factor of `4`.

**Comment by nickdos** [comment]
It looks like the `log_source_type_id` is being lost and `log_reason_type_id` is being read as `10` (testing) for all events -  by the trigger.

```
mysql> select * from event_summary_breakdown_reason_entity_source where month = '202103' limit 100;
+--------+-------------------+--------------------+------------+--------------------+------------------+--------------+
| month  | log_event_type_id | log_reason_type_id | entity_uid | log_source_type_id | number_of_events | record_count |
+--------+-------------------+--------------------+------------+--------------------+------------------+--------------+
| 202103 |              1002 |                 10 | co10       |                 -1 |                5 |         8664 |
| 202103 |              1002 |                 10 | co111      |                 -1 |                1 |            1 |
| 202103 |              1002 |                 10 | co117      |                 -1 |                5 |         7811 |
| 202103 |              1002 |                 10 | co118      |                 -1 |                1 |            1 |
```

and the last event shows:

```
mysql> select id,month,log_event_type_id,log_reason_type_id,log_source_type_id from log_event order by id desc limit 2;
+------------+--------+-------------------+--------------------+--------------------+
| id         | month  | log_event_type_id | log_reason_type_id | log_source_type_id |
+------------+--------+-------------------+--------------------+--------------------+
| 1047004135 | 202103 |              1002 |                 12 |                  1 |
| 1047004134 | 202103 |              1002 |                 13 |                  0 |
+------------+--------+-------------------+--------------------+--------------------+
```

So we should see `log_reason_type_id` values for 12 and 13 and `log_source_type_id` for 0 and 1.

Noting the[ trigger code](https://github.com/AtlasOfLivingAustralia/ala-install/blob/master/ansible/roles/logger-service/files/db/update_breakdown_summary_information.sql#L39) defaults to `-1` when a value can't be found.

**Comment by nickdos** [comment]
Looking at the `log_event_detail` table, it does not match either the `sourceBreakdown` nor `reasonBreakdown` output for `thisMonth` (but `reasonBreakdown` shows correct number of events but not records).

```
mysql> select * from event_summary_totals where month = '202103' limit 100;
+--------+-------------------+------------------+--------------+
| month  | log_event_type_id | number_of_events | record_count |
+--------+-------------------+------------------+--------------+
| 202103 |              1002 |               15 |       203800 |
+--------+-------------------+------------------+--------------+
```

versus `sourceBreakdown`:
```json
{
  "thisMonth": {
    "events": 510,
    "records": 397006
  }
}
```
and `reasonBreakdown`:
```json
{
  "thisMonth": {
    "events": 15,
    "records": 195499
  }
}
```
So the latter is only out by `8,301` records

**Comment by djtfmartin** [comment]
From what I can see the `/ws/sourceBreakdown` has never worked without an `entityUid` parameter.

This is because the only summary table that has source information is the `event_summary_breakdown_reason_entity_source`
and if an `entityUid` parameter isn't included then log events aggregates are incorrectly multiplied by the number of entities involved in the log event.

I suspect this line of [code](https://github.com/AtlasOfLivingAustralia/logger-service/blob/develop/grails-app/controllers/au/org/ala/logger/LoggerController.groovy#L156) suggests that `entityUid` was considered a mandatory parameter for the web service at some stage.

So I suggest we either:

* Add the check to the service to make `entityUid` mandatory

or

* Add additional support tables and queries to support the source breakdowns.

As its never worked (as far as I can see) i prefer the former.

**Comment by nickdos** [comment]
Checking `blame`, code has been there since Adam made a commit called [Add source type breakdown reporting](https://github.com/AtlasOfLivingAustralia/logger-service/commit/188c107210f5fe3d64b808a3b98fd1cf288f15cf).

**Comment by djtfmartin** [comment]
> Requiring entityUid would mean we couldn't look to see how downloads are being spread across the different source platforms, for all downloads (only for a given dataresource or collection), e.g. ALA4R vs ALA vs AVH for all downloads, which is quite useful - I'd argue more useful for us than just knowing how those spreads happen for a single entityUid (which is useful for the dataset custodians).

I agree it could be useful. I think we would need an `event_summary_breakdown_source` table (and possibly an `event_summary_breakdown_source_entity` table)  to support it and triggers to populate this, and changes to logger-service to query these new tables where an `entityUid` hasn't been supplied (which is what [happens with the reason breakdowns](https://github.com/AtlasOfLivingAustralia/logger-service/blob/develop/grails-app/services/au/org/ala/logger/LoggerService.groovy#L406)).

**Comment by nickdos** [comment]
Deployed to test but now home page link is broken due to requiring a `entityUid` param. So that needs fixing with an example `entityUid` like CSIRO (in6) or Aust Musium (in4). Moving back to in progress.

**Comment by Rita-C** [comment]
> Also, looking at code, there are more similar checks for `entityUid` required:
> 
>     * [ ]  https://github.com/AtlasOfLivingAustralia/logger-service/blob/develop/grails-app/controllers/au/org/ala/logger/LoggerController.groovy#L233

If no `entityUid` is supplied then:
https://github.com/AtlasOfLivingAustralia/logger-service/blob/develop/grails-app/services/au/org/ala/logger/LoggerService.groovy#L269

**Comment by nickdos** [comment]
Yes, that code needs changing too. Dave says the SP can't support looking up without an `entityUid` for reason breakdowns.

**Comment by Rita-C** [comment]
https://upsource.ala.org.au/logger-service/review/LS-CR-2

**Comment by nickdos** [comment]
Code reviewed ✅

**Comment by nickdos** [comment]
Testing this with the following endpoints shows it is returning the expected number of download events and record counts...

https://logger-test.ala.org.au/service/sourceBreakdown?eventId=1002&entityUid=in4
vs
https://logger-test.ala.org.au/service/reasonBreakdown?eventId=1002&entityUid=in4
